File: Clojure/Days/06/Find the odd int.clj
https://www.codewars.com/kata/54da5a58ea159efa38000836

; First solution
(ns find-the-odd-int)

(defn find-odd [xs]
  (->> xs 
    (group-by #(int %))
    (vals)
    (filter #(odd? (count %)))
    (apply first))  
)

; Using ffirst
(ns find-the-odd-int)

(defn find-odd [xs]
  (->> xs 
    (group-by #(int %))
    (vals)
    (filter #(odd? (count %)))
    (ffirst))  
)

; Using frequencies
(ns find-the-odd-int)

(defn find-odd [xs]
  (->> xs 
    (frequencies)
    (filter #(odd? (second %)))
    (ffirst)
    )  
)

; Magic from Clever guys
(ns find-the-odd-int)

(defn find-odd [xs]
  (reduce bit-xor xs))

File: F#/Day 03/Find the odd int.fs
// https://www.codewars.com/kata/54da5a58ea159efa38000836/

// First Solution

let inline findOdd ( numbers: int list) = 
    numbers
    |> List.groupBy (fun x -> x)
    |> List.find (fun (k, v) -> v.Length % 2 <> 0)
    |> fst

// Using id

let inline findOdd (numbers: int list) = 
    numbers
    |> List.groupBy id
    |> List.find (fun (_, v) -> v.Length % 2 <> 0)
    |> fst
